### PR TITLE
HIVE-22458: Add more constraints on showing partitions

### PR DESCRIPTION
**Pull request by dengzhhu653**
Add where/order by/limit constraints when showing partitions

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java [hunk 1]
```diff
@@ -57,10 +67,75 @@ public void analyzeInternal(ASTNode ast) throws SemanticException {
...
     ShowPartitionsDesc desc = new ShowPartitionsDesc(tableName, ctx.getResFile(), partSpec);
+    analyzeShowPartitionsConstraints(ast, table, desc);
     Task<DDLWork> task = TaskFactory.get(new DDLWork(getInputs(), getOutputs(), desc));
     rootTasks.add(task);
 
     task.setFetchSource(true);
     setFetchTask(createFetchTask(ShowPartitionsDesc.SCHEMA));
   }
+
+  private void analyzeShowPartitionsConstraints(ASTNode ast, Table tab,
```
Comment (miklosgergely) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java:78]: Please don't use long methods like this at ddl. Instead  cut it to 3 different methods, similar to getPartitionSpecs, each retrieving one data, returning it to analyzeInternal. Also please don't use setters at ShowPartitionDesc, a desc object at ddl should have final variables set at the constructor, and not by setters. This declares that they may not change later. Thanks.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionsOperation.java [hunk 2]
```diff
@@ -44,11 +60,49 @@ public int execute() throws HiveException {
       throw new HiveException(ErrorMsg.TABLE_NOT_PARTITIONED, desc.getTabName());
     }
 
-    List<String> parts = null;
-    if (desc.getPartSpec() != null) {
-      parts = context.getDb().getPartitionNames(tbl.getDbName(), tbl.getTableName(), desc.getPartSpec(), (short) -1);
+    List<String> parts;
+    if (desc.getCond() != null || desc.getOrder() != null) {
+      ExprNodeDesc pred = desc.getCond();
```
Comment (miklosgergely) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionsOperation.java:65]: This is long, please move it to a method.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java (hunk 1 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java]: @miklosgergely Thanks for reviewing. Update codes to move the parser into 3 different methods and make the variable in ShowPartitionDesc all final.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionsOperation.java (hunk 2 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionsOperation.java]: Move the logic into method getPartNamesByConstraints alone

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java [hunk 3]
```diff
@@ -89,7 +89,8 @@ public void analyzeInternal(ASTNode ast) throws SemanticException {
     setFetchTask(createFetchTask(ShowPartitionsDesc.SCHEMA));
   }
 
-  private ExprNodeDesc getShowPartitionsFilter(Table table, ASTNode command) throws SemanticException {
+
```
Comment (miklosgergely) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java:92]: Please add @VisibleForTesting

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/partition/show/ShowPartitionAnalyzer.java:93]: done

**Pull request closed without merging**